File: src/main/java/com/gcu/conversationservice/data/DataAccessInterface.java
package com.gcu.conversationservice.data;

import java.util.List;
import java.util.Optional;

public interface DataAccessInterface<T>{

    public List<T> findAll();

    public Optional<T> findByID(T t);

    /**
     * A method to create an object in the database
     * @param t The object to be persisted to the database
     * @return int The number of affected database rows
     */
    public Optional<T> create(T t);

    /**
     * Removes an object from the database
     * @param t The object to be removed
     * @return int The number of affected database rows
     */
    public boolean delete(T t);
}

File: src/main/kotlin/com/gcu/userservice/data/DataAccessInterface.kt
package com.gcu.userservice.data

interface DataAccessInterface<T> {

    fun findAll(): List<T>
    fun findAllWithID(id: String): List<T>
    fun findByID(id: String): T?
    fun findBy(t: T): T?
    fun findByString(search: String): T?
    fun create(t: T): Boolean
    fun update(t: T): Boolean
    fun delete(t: T): Boolean
}